package/MDAnalysis/lib/distances.py

```diff
@@ -608,6 +609,61 @@ def _pkdtree_capped(reference, configuration, max_cutoff, min_cutoff=None, box=N
                 distances.append(dist[num])
     return pairs, distances
 
+def _nsgrid_capped(reference, configuration, max_cutoff, min_cutoff=None,
+                   box=None):
+    """Search all the pairs in *reference* and *configuration* within 
+    a specified distance using Grid Search
+
+    
+    Parameters
+    -----------
+    reference : array
+        reference coordinates array with shape ``reference.shape = (3,)``
+        or ``reference.shape = (len(reference), 3)``.
+    configuration : array
+        Configuration coordinate array with shape ``reference.shape = (3,)``
+        or ``reference.shape = (len(reference), 3)``
+    max_cutoff : float
+        Maximum cutoff distance between the reference and configuration
+    min_cutoff : (optional) float
+        Minimum cutoff distance between reference and configuration [None]
+    box :  array 
+        The dimensions, if provided, must be provided in the same
+        The unitcell dimesions for this system format as returned
+        by :attr:`MDAnalysis.coordinates.base.Timestep.dimensions`:
+        ``[lx,ly, lz, alpha, beta, gamma]``. Minimum image convention
+        is applied if the box is provided 
+
+    Note
+    ----
+    Non Periodic Boundary conditions can-not be handled currently
+
+    """
+    from .nsgrid import FastNS
+    if reference.shape == (3, ):
+        reference = reference[None, :]
+    if configuration.shape == (3, ):
+        configuration = configuration[None, :]
+    all_coords = np.concatenate([configuration, reference])
+    mapping = np.arange(len(reference), dtype=np.int64)
+    gridsearch = FastNS(box, max_cutoff, all_coords)
+    gridsearch.prepare()
+    search_ids = np.arange(len(configuration), len(all_coords))
+    results = gridsearch.search(search_ids=search_ids)
+    pairs = results.get_pairs()
+    pairs[:, 1] = undo_augment(pairs[:, 1], mapping, len(configuration))
+    pair_distance = results.get_pair_distances()
+    
+    if min_cutoff is not None:
+        idx = pair_distance > min_cutoff
+        pairs, pair_distance = pairs[idx], pair_distance[idx]
+    if pairs.size > 0:
```

Comment:
@ayushsuhane I don't really understand what this step is doing?

Reply:
The problem with concatenate and search id is that it wouldn't mask the pairs within search ids.  The pairs which are returned from capped function are such that in a pair ``(i, j)`` i is from the reference positions and j is from the configuration positions. But here, it will return all the indices i.e ``(i, j)`` might be such that i and j both are indices of the search group.

Reply:
Ok so it sounds like this is much better for self_capped_distance than capped_distance right now.  Ideally we'd create 2 nsgrids and tell them to search each other?